src/dmd/expressionsem.d

```diff
@@ -5834,33 +5821,11 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }()
             */
 
-            /*
-            Stores the result of an operand expression into a temporary
-            if necessary, e.g. if it is an impure fuction call containing side
-            effects as in https://issues.dlang.org/show_bug.cgi?id=20114
-
-            Params:
-                op = an expression which may require a temporary and will be
-                     replaced by `(auto tmp = op, tmp)` if necessary
-
-            Returns: `op` or `tmp` for subsequent access to the possibly promoted operand
-            */
-            Expression maybePromoteToTmp(ref Expression op)
+            // if the assert condition is a mixin expression, try to compile it
+            if (auto ce = exp.e1.isCompileExp())
             {
-                if (op.hasSideEffect)
-                {
-                    const stc = STC.exptemp | (op.isLvalue() ? STC.ref_ : STC.rvalue);
-                    auto tmp = copyToTemp(stc, "__assertOp", op);
-                    tmp.dsymbolSemantic(sc);
-
-                    auto decl = new DeclarationExp(op.loc, tmp);
-                    auto var = new VarExp(op.loc, tmp);
-                    auto comb = Expression.combine(decl, var);
-                    op = comb.expressionSemantic(sc);
-
-                    return var;
-                }
-                return op;
+                if (auto e1 = compileIt(ce))
```

Comment:
Shouldn't this return if the `mixin` failed to compile?

Reply:
It'll be compiled again during sema of the final condition (full `CompileExp` sema), which takes care of error promotion.